src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X509Certificate2.cs

```diff
@@ -1166,6 +1167,239 @@ public bool TryExportCertificatePem(Span<char> destination, out int charsWritten
+        ///       </item>
+        ///       <item>
+        ///         If there is no Subject Alternative Name extension, or the extension does not have any entries
+        ///         of the appropriate type, then Common Name matching is used as a fallback.
+        ///       </item>
+        ///       <item>
+        ///         For Common Name matching, if the Subject Name contains a single Common Name, and that attribute
+        ///         is not defined as part of a multi-valued Relative Distinguished Name, then the hostname is matched
+        ///         against the Common Name attribute's value.
+        ///         Note that wildcards are not used in Common Name matching.
+        ///       </item>
+        ///     </list>
+        ///   </para>
+        ///   <para>
+        ///     This method does not convert non-ASCII hostnames to the IDNA representation. For Unicode domains,
+        ///     the caller must make use of <see cref="System.Globalization.IdnMapping"/> or an equivalent IDNA mapper.
+        ///   </para>
+        ///   <para>
+        ///     The "exact" matches performed by this routine are <see cref="StringComparison.OrdinalIgnoreCase"/>,
+        ///     as domain names are not case-sensitive.
+        ///   </para>
+        ///   <para>
+        ///     This method does not determine if the hostname is authorized by a trusted authority.  A trust
+        ///     decision cannot be made without additionally checking for trust via <see cref="X509Chain"/>.
+        ///   </para>
+        ///   <para>
+        ///     This method does not check that the certificate has an <c>id-kp-serverAuth</c> (1.3.6.1.5.5.7.3.1)
+        ///     extended key usage.
+        ///   </para>
+        /// </remarks>
+        /// <exception cref="CryptographicException">
+        ///   <para>The certificate contains multiple Subject Alternative Name extensions.</para>
+        ///   <para>- or -</para>
+        ///   <para>The Subject Alternative Name extension or Subject Name could not be decooded.</para>
+        /// </exception>
+        public bool MatchesHostname(string hostname, bool allowWildcards = true, bool allowCommonName = true)
+        {
+            ArgumentNullException.ThrowIfNull(hostname);
+
+            if (hostname.Length == 0)
+            {
+                return false;
+            }
+
+            X509Extension? rawSAN = null;
+
+            foreach (X509Extension extension in Pal.Extensions)
+            {
+                if (extension.Oid!.Value == Oids.SubjectAltName)
+                {
+                    if (rawSAN is null)
+                    {
+                        rawSAN = extension;
+                    }
+                    else
+                    {
+                        throw new CryptographicException(SR.Cryptography_X509_TooManySANs);
+                    }
+                }
+            }
+
+            if (rawSAN is not null)
```

Comment:
Is there a chance this will return a strongly typed `X509SubjectAlternativeNameExtension`? Could this do an `is X509SubjectAlternativeNameExtension` check before creating and copying in to a new instance?

Reply:
No, `Pal.Extensions` always returns untyped extensions, then X509Certificate2.Extension creates a stronger type, if applicable, and uses CopyFrom to load them in.

Added an assert, though.

Reply:
Ah, okay, I glossed over `Pal.` in front of Extensions.